sapio-psbt/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+    let buf = tokio::fs::read(file_name).await?;
+    Ok(ExtendedPrivKey::decode(&buf)?)
+}
+
+pub fn sign(
+    xpriv: ExtendedPrivKey,
+    mut psbt: PartiallySignedTransaction,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<Vec<u8>, PSBTSigningError> {
+    sign_psbt_inplace(&xpriv, &mut psbt, &Secp256k1::new(), hash_ty)?;
+    let bytes = serialize(&psbt);
+    Ok(bytes)
+}
+pub async fn show_pubkey(input: &std::ffi::OsStr) -> Result<(), Box<dyn Error>> {
+    let buf = tokio::fs::read(input).await?;
+    let xpriv = ExtendedPrivKey::decode(&buf)?;
+    println!("{}", ExtendedPubKey::from_priv(&Secp256k1::new(), &xpriv));
+    Ok(())
+}
+pub fn new_key(network: &str, out: &std::ffi::OsStr) -> Result<(), Box<dyn Error>> {
+    let entropy: [u8; 32] = rand::thread_rng().gen();
+    let xpriv = ExtendedPrivKey::new_master(Network::from_str(network)?, &entropy)?;
+    std::fs::write(out, &xpriv.encode())?;
+    println!("{}", ExtendedPubKey::from_priv(&Secp256k1::new(), &xpriv));
+    Ok(())
+}
+
+pub fn sign_psbt_inplace(
+    xpriv: &ExtendedPrivKey,
+    psbt: &mut PartiallySignedTransaction,
+    secp: &Secp256k1<All>,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<(), PSBTSigningError> {
+    sign_psbt_input_inplace(xpriv, psbt, secp, 0, hash_ty)
+}
+
+#[derive(Debug, Clone)]
+pub enum PSBTSigningError {
+    NoUTXOAtIndex(usize),
+    NoInputAtIndex(usize),
+}
+
+impl Display for PSBTSigningError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+impl Error for PSBTSigningError {}
+
+pub fn sign_psbt_input_inplace(
+    xpriv: &ExtendedPrivKey,
+    psbt: &mut PartiallySignedTransaction,
+    secp: &Secp256k1<All>,
+    idx: usize,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<(), PSBTSigningError> {
+    let tx = psbt.clone().extract_tx();
+    let utxos: Vec<TxOut> = psbt
+        .inputs
+        .iter()
+        .enumerate()
+        .map(|(i, o)| {
+            if let Some(ref utxo) = o.witness_utxo {
+                Ok(utxo.clone())
+            } else {
+                Err(i)
+            }
+        })
+        .collect::<Result<Vec<TxOut>, usize>>()
+        .map_err(|u| PSBTSigningError::NoUTXOAtIndex(u))?;
+    let mut sighash = bitcoin::util::sighash::SighashCache::new(&tx);
+    let input = &mut psbt
+        .inputs
+        .get_mut(idx)
+        .ok_or(PSBTSigningError::NoInputAtIndex(idx))?;
+    let prevouts = &Prevouts::All(&utxos);
+    sign_taproot_top_key(xpriv, secp, input, &mut sighash, prevouts, hash_ty);
+    sign_all_tapleaf_branches(xpriv, secp, input, &mut sighash, prevouts, hash_ty);
+    Ok(())
+}
+
+fn sign_all_tapleaf_branches(
+    xpriv: &ExtendedPrivKey,
+    secp: &Secp256k1<All>,
+    input: &mut bitcoin::psbt::Input,
+    sighash: &mut bitcoin::util::sighash::SighashCache<&bitcoin::Transaction>,
+    prevouts: &Prevouts<TxOut>,
+    hash_ty: bitcoin::SchnorrSighashType,
+) {
```

Comment:
In line with my previous comment it may be good to return the input from this function rather than mutating in place, the cache seems appropriate to mutate though, since it is context that is not part of the main operation.